src/test_runner.cr

```diff
@@ -43,18 +43,44 @@ module Mint
                   }
                 }
 
-                this.suites = suites
+                let error = null;
+
+                window.onerror = (message) => {
+                  if (this.socket.readyState === 1) {
+                    this.crash(message);
+                  } else {
+                    error = error || message;
+                  }
+                }
 
                 this.socket.onopen = () => {
-                  this.run()
-                    .then(() => this.socket.send("DONE"))
+                  if (error != null) {
+                    this.crash(error);
+                  }
+
+                  window.addEventListener('unhandledrejection', (event) => {
+                    event.promise.catch(e => this.crash(e.toString()));
+                  });
```

Comment:
This I think should be moved out of `onopen` handler as it has nothing to do with it - could be put just right next to `window.onerror = ...` assignment.

Reply:
Yes, that's possible, but since "crashing" means sending the `CRASHED` message through the socket, I purposefully placed it here since it's not possible for any unhandled rejections to occur until the tests have started (at which point the socket is already open and ready to receive messages). This way, it avoids the need to "cache" the error in a variable (like in the `window.onerror` handler, which catches errors that may occur before the socket is open, such as syntax errors).

Reply:
I think we should avoid binding `unhandledrejection` - it may propagate errors which might be acceptable and shouldn't crash the runner - logging 'em might be useful though.

Reply:
Any unhandled exception that can reach this handler will always interrupt the test execution; we can log the errors, but the test runner would still hang forever. Alternatively, if you know the specific exceptions that are harmless, we can create a passthrough list that will make the handler skip calling `this.crash` if the exception in question is in the list.

Reply:
Aren't you confusing unhandled (promise) _rejection_ with unhandled _exception_?

Reply:
We could need both since `onerror` doesn't catch unhandled promises. 
https://codepen.io/gdotdesign/pen/oNYXENp?editors=1010

I'm not sure that we should bind it to the runner or not, theoretically there are cases where it could fail to catch rejections for example in a stores initial state, but that wouldn't cause a  crash.

> I think we should avoid binding `unhandledrejection` - it may propagate errors which might be acceptable and shouldn't crash the runner - logging 'em might be useful though.

I don't think there are system promises that it can catch, but in any case we can log them.

Reply:
If I understood correctly, [this page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch) shows that exceptions being thrown in a Promise context will result in a rejection, which, if unhandled by a `.catch` chained function, will propagate to this global handler (i.e. an unhandled exception in the Promise becomes an unhandled rejection when propagated outside).

Actually, I just realized that, since we now have a `finally` chained function, it will guarantee that the test runner shuts down even if an error occurs, which means that this handler no longer needs to crash the test runner anymore. I'll change it to just log the error instead. The `onerror` handler still needs to crash the test runner, though, since reaching it would mean that something serious (like a syntax error) has occurred.